src/auth/src/credentials/external_account_sources/executable_sourced.rs

```diff
@@ -0,0 +1,317 @@
+            ExecutableResponse {
+                message: Some(message),
+                code: Some(code),
+                ..
+            } => {
+                let msg =
+                    format!("{MSG}, response contains unsuccessful response: ({code}) {message}");
+                CredentialsError::from_msg(false, msg)
+            }
+            _ => {
+                let msg = format!(
+                    "{MSG}, response must include `code` and `message` fields when unsuccessful"
+                );
+                CredentialsError::from_msg(false, msg)
+            }
+        }
+    }
+}
+
+const MSG: &str = "failed to read subject token";
+const DEFAULT_TIMEOUT_SECS: u32 = 30;
+const ALLOW_EXECUTABLE_ENV: &str = "GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES";
+
+#[async_trait::async_trait]
+impl SubjectTokenProvider for ExecutableSourcedCredentials {
+    async fn subject_token(&self) -> Result<String> {
+        let output = match self.executable.clone() {
+            ExecutableConfig {
+                output_file: Some(output_file),
+                ..
+            } => Self::from_output_file(output_file).await,
+            ExecutableConfig {
+                command: Some(command),
+                timeout_millis,
+                ..
+            } => {
+                let timeout = match timeout_millis {
+                    Some(timeout) => Duration::from_millis(timeout.into()),
+                    None => Duration::from_secs(DEFAULT_TIMEOUT_SECS.into()),
+                };
+                Self::from_command(command, timeout).await
+            }
+            _ => Err(CredentialsError::from_msg(
+                false,
+                format!("{MSG}, either `output_file` or `command` needs to be informed"),
+            )),
+        }?;
+
+        let output = output.trim().to_string();
+        let subject_token = Self::parse_token(output)?;
+
+        if subject_token.is_empty() {
+            let msg = format!("{MSG}, subject token is empty");
+            return Err(CredentialsError::from_msg(false, msg));
+        }
+
+        Ok(subject_token)
+    }
+}
+
+impl ExecutableSourcedCredentials {
+    async fn from_output_file(output_file: String) -> Result<String> {
+        let content = std::fs::read_to_string(output_file)
+            .map_err(|e| CredentialsError::from_source(false, e))?;
+        Ok(content)
+    }
+
+    async fn from_command(command: String, timeout: Duration) -> Result<String> {
+        // For security reasons, we need our consumers to set this environment variable to allow executables to be run.
+        let allow_executable = std::env::var(ALLOW_EXECUTABLE_ENV)
+            .ok()
+            .unwrap_or("0".to_string());
+        if allow_executable != "1" {
+            return Err(CredentialsError::from_msg(
+                false,
+                "executables need to be explicitly allowed (set GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES to '1') to run",
+            ));
+        }
```

Comment:
This is very strange.  So they set an explicit configuration to use an executable and *then* they also need to set an environment variable?

Reply:
yeah, I agree that it's weird, but out of my control as I'm just following the AIP guide.

Reply:
@lsirac can you answer this

Reply:
This is for security reasons e.g. considering SaaS vendors and shared machines. Some SaaS vendors allow uploading JSON key files. 

Malicious actors can exploit the file format of 3PI credentials to access resources on the shared machines running the SaaS software that should otherwise be inaccessible.